### DISCUSSION TITLE
New - Grid Row-Gap

### DISCUSSION THREAD

**Discussion by markblandford** [Legal-and-General/canopy#1046] [Canopy requests]
**Type of request**
New

**Component, pattern or foundation?**
Foundation maybe?

**Nature of request**

Unless I'm mistaken (I can't find anything in the docs of Canopy or Flexboxgrid), there is no built-in way to specify the CSS [`row-gap`](https://developer.mozilla.org/en-US/docs/Web/CSS/row-gap) property on an `lgRow` element. As such, when a row 'stacks' there is no Canopy way to specify the 'gap' between flex items in the row.

For example, this HTML:

```html
<div lgContainer>
  <div lgRow [style.margin-bottom]="'1rem'">
    <div lgColLg="4" lgCol="12">
      <lg-card lgMarginBottom="none" [style.height]="'100%'">
        <lg-card-content>
          <lg-heading level="2.0">Card 0</lg-heading>
          <p>
            Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod
            tempor incididunt ut labore et dolore magna aliqua.
          </p>
        </lg-card-content>
      </lg-card>
    </div>
    <div [lgColLg]="4" lgCol="12">
      <lg-card lgMarginBottom="none" [style.height]="'100%'">
        <lg-card-content>
          <lg-heading level="2.0">Card 1</lg-heading>
          <p>
            Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut
            aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in
            voluptate velit esse cillum dolore eu fugiat nulla pariatur.
          </p>
        </lg-card-content>
      </lg-card>
    </div>
  </div>
</div>
```

Generates:

| LG Viewport | SM Viewport |
|--|--|
| ![image](https://user-images.githubusercontent.com/8008207/194500931-b5c51ac0-8210-4fa9-bdd5-7f8e5861fd5f.png) | ![image](https://user-images.githubusercontent.com/8008207/194501038-9b602c89-42d7-48be-8ad8-9b845edb32bc.png) |

As you can see, there is no 'gap' between the flex items when stacked (I've removed the margin on the cards for demonstration purposes only).

Adding the `row-gap: 1rem;` to the `lgRow` `<div>`, leads to no change in the rendering when the cards / flex items are in a row, but when they become stacked, we see a nice, consistent `1rem` gap between each item:

![image](https://user-images.githubusercontent.com/8008207/194501851-39effccc-2d3d-4c9d-85d0-514ca9c8c3c4.png)

**Proposal**

I propose the following:

1. A new `lgRowGap=` directive, where a consumer of Canopy can specify the CSS row-gap property and it's value. If a value is not provided, it will default to `var(--space-sm)`.
2. New CSS classes, `.lg-row-gap--#{$val}` along the same lines as as the existing, `.lg-margin--#{$val}`, so the same can be achieved with a CSS class.
3. If neither the new directive or CSS classes are provided, the existing behaviour is observed.

_I am not proposing any `column-gap` behaviour, as I feel the existing Canopy grid sort of already deals with this quite nicely (and was previously discussed in [issue 198 in Flexboxgrid](https://github.com/kristoferjoseph/flexboxgrid/issues/198))._

This can of course be achieved, as I have done above, by simply adding the style in the consuming app, but given the 'grid' is handled by Canopy, I feel it should be able to support this.

If this is accepted, I would like the opportunity to contribute and create a PR.

**Comment by alexcanning** [comment]
Hi Mark, thanks for raising a request. Maybe I've misunderstood but are you saying that you're seeing zero margin by default? Or that you'd like the ability to specify a custom margin?

@elenagarrone I wonder if this is a bug? There's definitely no scenario where we'd deliberately want zero horizontal margin between cards.

**Reply by markblandford** [reply]
I don't think it is a bug per-say, I think it is by design in flexboxgrid (which from the docs is what the Canopy grid is based upon). Perhaps I'm confusing things a little by mentioning cards.

If I change everything to `<div>` elements, an `lgRow` is nicely spaced when horizontal:

(purple lines are the Firefox dev tools showing the flexbox items)

![image](https://user-images.githubusercontent.com/8008207/194514034-d4b0b65b-3356-453a-9648-596c6229a0b5.png)

No code changes (so no `row-gap` defined), on a mobile device, this same `lgRow` stacks as:

![image](https://user-images.githubusercontent.com/8008207/194513958-c355399f-a295-4bcd-89f5-61191c8bdc06.png)

**Reply by markblandford** [reply]
From what I can make out in the code, padding left/right is catered for with the grid scss (in the `col-style` mixin). However, there is nothing for vertical spacing.

**Reply by elenagarrone** [reply]
@markblandford Thanks, I think it would be a good idea to add the `row-gap` property as you mentioned. I think up until now we have always used the margin directives within the applications for it but I think this property is a better solution. We might want to mark it as a breaking change so that we make sure we don't break the current apps. 

@paul-request what do you think? We could merge this change into `next` so that it will be merged together with the work you are doing on the header nav?

**Reply by paul-request** [reply]
@elenagarrone The problem with next is it would have to wait until I have finished the rest of the nav, and would have to perhaps update the apps as the dimensions of the header have changed a bit (might not have to if we are all using `lg-page` correctly?

**Reply by markblandford** [reply]
@elenagarrone @paul-request if this is a new directive & CSS classes, would this be a breaking change if the original behaviour isn't impacted?

**Reply by elenagarrone** [reply]
@markblandford it's only a breaking change if we add the default value to the `lgRow` directive in my opinion as other apps have other methods of adding the spacing between cards on small devices. The new `lgRowGap` directive alone is not.

We could maybe add the `lgRowGap` directive without adding the default value to `lgRow` so that the change can be merged into `master`. Then we could create a separate PR to go into `next` to add the default as a breaking change.

**Reply by markblandford** [reply]
I'll write some code, raise a draft PR and we can then take it from there. Thank you everyone!

**Reply by markblandford** [reply]
@elenagarrone @paul-request I've created a draft PR here: #1052

**Reply by elenagarrone** [reply]
This is now merged and available in [v9.1.0](https://github.com/Legal-and-General/canopy/releases/tag/v9.1.0). Thanks @markblandford for contributing 👍
